### ISSUE TITLE
Question about voxel coordinates

### ISSUE THREAD

**Issue by woxihuanjiangguo** [open-mmlab/mmdetection3d#1796]
In the MvxTwoStage Detector, there is this method voxelize:
`        def voxelize(self, points):
        """Apply dynamic voxelization to points.

        Args:
            points (list[torch.Tensor]): Points of each sample.

        Returns:
            tuple[torch.Tensor]: Concatenated points, number of points
                per voxel, and coordinates.
        """
        voxels, coors, num_points = [], [], []
        for res in points:
            res_voxels, res_coors, res_num_points = self.pts_voxel_layer(res)
            voxels.append(res_voxels)
            coors.append(res_coors)
            num_points.append(res_num_points)
        voxels = torch.cat(voxels, dim=0)
        num_points = torch.cat(num_points, dim=0)
        coors_batch = []
        for i, coor in enumerate(coors):
            coor_pad = F.pad(coor, (1, 0), mode='constant', value=i)
            coors_batch.append(coor_pad)
        coors_batch = torch.cat(coors_batch, dim=0)
        return voxels, num_points, coors_batch
`

that returns coors_batch. Its shape is N*4, while the "4" dimension represents (batch_idx, z, y, x), if I'm understanding this correctly. I notice that z, y, x are all positive integers.

Is this (z, y, x) coherent with the lidar coordinates? For coherence, I mean is the center of the voxel grid also the center of the lidar coordinate, and are their x, y, z axes parallel with each other?

**Comment by Tai-Wang** [comment]
Not exactly. The coordinates are used to locate a specific voxel in a voxel feature map, so its origin should be moved to a corner. You can find the derivation details [here](https://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/csrc/pytorch/cuda/voxelization_cuda.cu#L32).

**Comment by woxihuanjiangguo** [comment]
Thank you! I got it now.
